Added some API skeleton, db schema, mocks and stubs

kvoronov opened the pull request.

kvoronov committed:
ci test

kvoronov committed:
Merge branch 'development' of https://github.com/VolunteerService/uvs into development

orn0t commented on app.js:
```diff
@@ -1,16 +1,61 @@
+    db = require('./db.js')(pg, prettify),
+    Session = require('./entities/Session.js')(_, pg, db),
+    User = require('./entities/User.js')(Session, pg, db),
+    Event = require('./entities/Event.js')(pg, db),
```
`Event` is reserved for js builtin events interface,  maybe use `model` prefix or suffix

kvoronov replied:
Окей.

orn0t commented on app.js:
```diff
@@ -1,16 +1,61 @@
-var app = express();
-
-mongoose.connect(process.env.MONGODB_URI);
+mongoose.connect(process.env.MONGODB_URI || "mongo://uvs@localhost");
```
Dont commit local connect strings or credentials

kvoronov replied:
Предлагается настраивать локальное окружение идентичным образом. Или могу отдельно механизм локальных конфигов намутить.

orn0t replied:
use environment variables, most process runners like foreman, pm2 support env configs.

orn0t commented on db.js:
```diff
@@ -0,0 +1,24 @@
+module.exports = function(pg, prettify){
+	var conectionString = process.env.DATABASE_URL || "postgres://postgres:123456@localhost/uvs";
```
Dont commit local connect strings or credentials

kvoronov replied:
Предлагается настраивать локальное окружение идентичным образом. Или могу отдельно механизм локальных конфигов намутить.

orn0t replied:
use environment variables, most process runners like foreman, pm2 support env configs.

orn0t commented on app.js:
```diff
@@ -1,16 +1,61 @@
+        } else {
+            if(!_.isObject(methods[session.userPriviledge]) || !_.isFunction(methods[session.userPriviledge][methodName])){
+                res.status(400).json({err: {
+                    code: 2,
```
Avoid magic numbers like`code: 2`

kvoronov replied:
Это внутренние коды ошибок. Можно будет держать их перечень в отдельной статье в вики.

orn0t replied:
by the way, if you want a constant - use it like a constant, not an unnamed integer.

orn0t commented on app.js:
```diff
@@ -1,16 +1,61 @@
+            if(!_.isObject(methods[session.userPriviledge]) || !_.isFunction(methods[session.userPriviledge][methodName])){
+                res.status(400).json({err: {
+                    code: 2,
+                    message: 'Access denied'
```
Use proper HTTP codes,  `403` for `Access denied`  not `400`

kvoronov replied:
В разных API (в том числе крупных вендоров) используются разные подмножества кодов HTTP, при этом видится нормальным использовать два кода (ну или 3 - 200, 400, 500) для post-запросов, являющихся вызовами методов API - чтобы отличать просто ошибку от не ошибки, а уже ошибки уровня нашей системы разруливать внутренними кодами. Это удобно впоследствии для обеспечения единой языковой среды и бесшовной интеграции с фронтом. Однако это не так принципиально, могу и присылать разные коды. А ошибка/не ошибка различать по 200/не 200.

orn0t replied:
Large vendors had a lot of legacy, inner protocol restrictions, and resources to manage, maintain and support their own standards. If we can fit with public standards like [HTTP 1.1 RFC7231](https://tools.ietf.org/html/rfc7231) we must.

orn0t commented on entities/Donate.js:
```diff
@@ -0,0 +1,21 @@
+
+	Donate = {
+		getMonthBudget: function(data, callback){
+            db.query('select * from patient ', [], function(err, result){
```
`patient` ?

kvoronov replied:
Это стабы, скопировано из старого проекта, вечером закину уже честных имплементаций.

orn0t replied:
Do not send PR with something inappropriate, review time costs.

orn0t commented on entities/Event.js:
```diff
@@ -0,0 +1,65 @@
+    // TODO: make honest queries to right db
+	Event = {
+		getAllEvents: function(data, callback){
+    		db.query('select * from patient ', [], function(err, result){
```
`patient` ?

kvoronov replied:
Это стабы, скопировано из старого проекта, вечером закину уже честных имплементаций.

orn0t commented on entities/Event.js:
```diff
@@ -0,0 +1,65 @@
+            })
+    	},
+    	editEvent: function(data, callback){
+            db.query(`update patient 
```
`patient` ?

kvoronov replied:
Это стабы, скопировано из старого проекта, вечером закину уже честных имплементаций.

orn0t commented on entities/Event.js:
```diff
@@ -0,0 +1,65 @@
+            })
+    	},
+    	getEventById: function(data, callback){
+            db.query('select * from patient where id = $1', [data.patientId], function(err, result){
```
`patient` ? Also SQLInjection here.

kvoronov replied:
Это стабы, скопировано из старого проекта, вечером закину уже честных имплементаций. Параметры экранируются. Модуль pg этим занимается.

orn0t commented on entities/Event.js:
```diff
@@ -0,0 +1,65 @@
+    	},
+
+		addEvent: function(data, callback){
+            db.query('insert into patient (firstname, lastname, weight, height, sex) values ($1, $2, $3, $4, $5)', 
```
`patient` ? Also SQLinjection here.

kvoronov replied:
Это стабы, скопировано из старого проекта, вечером закину уже честных имплементаций. Параметры экранируются. Модуль pg этим занимается.

orn0t commented on entities/Event.js:
```diff
@@ -0,0 +1,65 @@
+    	},
+
+		deleteEvent: function(data, callback){
+            db.query('delete from patient where id=$1', [data.patientId], function(err, result){
```
`patient` ? Also SQLinjection here.

kvoronov replied:
Это стабы, скопировано из старого проекта, вечером закину уже честных имплементаций. Параметры экранируются. Модуль pg этим занимается.

orn0t commented on entities/Project.js:
```diff
@@ -0,0 +1,64 @@
+
+	Project = {
+		getAllProjects: function(data, callback){
+    		db.query('select * from patient ', [], function(err, result){
```
`patient` ?

kvoronov replied:
Это стабы, скопировано из старого проекта, вечером закину уже честных имплементаций.

orn0t commented on entities/Project.js:
```diff
@@ -0,0 +1,64 @@
+            })
+    	},
+    	editProject: function(data, callback){
+            db.query(`update patient 
```
`patient` ? Also SQLinjection here.

kvoronov replied:
Это стабы, скопировано из старого проекта, вечером закину уже честных имплементаций. Параметры экранируются. Модуль pg этим занимается.

orn0t commented on entities/Project.js:
```diff
@@ -0,0 +1,64 @@
+    	},
+    	getProjectById: function(data, callback){
+            db.query('select * from patient where id = $1', [data.patientId], function(err, result){
+                if (err) {
```
`patient` ?

kvoronov replied:
Это стабы, скопировано из старого проекта, вечером закину уже честных имплементаций.

orn0t commented on entities/Session.js:
```diff
@@ -0,0 +1,69 @@
+        return text;
+    }
+
+    function createSession(data, callback){
```
Please use any popular open source `session` middleware, a.e `express-session` with proper db-backend.
We don't want to maintain additional code that can be replaced with open source analogue.

kvoronov replied:
Тянуть лишние зависимости ради 10 строк кода - не очень хорошо. Например, можно вспомнить модуль leftpad, который порушил пол npm'а, в том числе реакт, например. А это был модуль-однострочник.

orn0t replied:
each custom 10 lines costs 1h (30$) per contributor to lern/maintain/update,  open source analogues costs nothing.

kvoronov replied:
Масса опыта показывает, что вместе с готовым опенсорс-решением зачастую получаешь чужие баги. А кода там может быть строк 600 вместо 10. Мейнтейнить 10 строк своего кода лучше, чем 600 строк чужого.
А вообще, каждые 10 строк нашей баталии стоят "1h (30$) per contributor", так что мы тут уже баксов 500 проболтали :)

orn0t replied:
yep, and will spent more, but we will not write own session storage as same as we will not write own framework or own FB api wrapper because using an existed software will save ours and next another developers time and money

orn0t replied:
we wont write code just because we can, we must write only things we need and we cant simply get it from somewhere else. that the point

kvoronov replied:
Used express-session, dependent code rewrited. All working nice.

orn0t commented on entities/User.js:
```diff
@@ -0,0 +1,67 @@
+	var User = {
+        login: function(data, callback){
+            db.query(`select id from public.user 
+            where login = $1 and password = $2`, [data.login, data.password], function(err, result){
```
So, we store `raw` password strings?

kvoronov replied:
Нет, позже запилю bcrypt

orn0t replied:
ok, so place @todo here

orn0t commented on package.json:
```diff
@@ -18,7 +18,13 @@
+    "body-parser": "^1.15.2",
+    "cookie-parser": "^1.4.3",
+    "formidable": "^1.0.17",
+    "mongojs": "^2.4.0",
```
we already use `mongoose`, why we need `mongojs` dep?

kvoronov replied:
потому что это быстрый драйвер для прямого доступа, а модели mongoose - избыточность и бюрократия, если речь например о банальном логировании. Хотим строгие модели - для этого есть SQL, зачем тащить их в монгу?

orn0t replied:
Ok, lets use deps by `needs` not by `wants` or `futures`. Is any code usage for `mongojs` and `formidable` for now ?

orn0t commented on prettify.js:
```diff
@@ -0,0 +1,33 @@
+		participantType: 'participantType',
+		projectId: 'projectId',
+		userid: 'userId'
+	};
```
Hardcoded fields?

kvoronov replied:
См. комментарий по поводу функции prettify и принципов её работы

orn0t replied:
ok

orn0t commented on db.js:
```diff
@@ -0,0 +1,24 @@
+	        		console.error('error running query', err);
+	        	}
+	        	if (callback) {
+	        		result.rows =  prettify(result.rows) // e
```
You applying `prettify` to almost all queries, but it has only several fields hardcoded to work with. Looks like it wont work in general. You need to add test to prove it works.

kvoronov replied:
При добавлении новых полей в базу, если им требуется перевод в lowerCamelCase (то есть, если в названии поля более одного слова) - добавляется одна строка в массив. Для большинства полей это не требуется. Функция prettify нужна, чтобы, будучи адаптером к постгресу, решить проблему того, что постгрес не различает регистр букв и возвращает все названия полей в lowercase, а работать с ними в таком виде неудобно и не красиво.

orn0t replied:
Can we apply KISS here and avoid those `sugar` needs ?

kvoronov replied:
Ну, если это никак специально не постобрабатывать, то на фронт могут возвращаться объекты с полями наподобие volunteersquota вместо volunteersQuota, что не очень читабельно. А если сделать в базе snake_case для полей, то в постобработке для перевода в lowerCamelCase будут нуждаться все поля. Правда, в таком случае метод будет работать на любых именах полей, но он всё равно будет необходим.

orn0t replied:
ok, show me the code we have that will not work or will look ambiguous without this?

orn0t replied:
we dont need code just because it cool or because someone already made it, we need code that works, that useful, and that is simplest as possible.

kvoronov replied:
All code at frontend and backend, which work with these fields will look ugly when accessing object fields like obj.volunteersquota or obj.participanttype, etc.
We lose some information about field names when they are lowercased. So we restoring that lost info such way.

orn0t reviewed (changes requested):
1. Use `Rebase` not `Merge` strategy  to avoid local merge commits in repo
2. Use feature-branches to avoid historical commits in PR 
3. Fix newlines at the end of files
4. How to manage DB changes? We cant drop/import DB on each change by hands.

orn0t reviewed (changes requested):
1. Fix newlines at the end of files
2. How do we manage db ? 
3. Squash paltry commits 
4. Update code, its outdated now

kvoronov committed:
Added API and db skeleton and some stubs
Added password hashing using pure js module bcrypt-nodejs (native module bcrypt not used due to failing of building with node-gyp on heroku).
Added session and user roles mechanism using express-session

kvoronov commented:
Squashed all commits to one. Used express-session instead of own session solution.
Create db applying schema from uvs_postgres_create.sql
Later manage applying migration patches
